core/src/main/java/io/questdb/griffin/engine/join/NestedLoopLeftJoinRecordCursorFactory.java

```diff
@@ -0,0 +1,173 @@
+        RecordCursor masterCursor = masterFactory.getCursor(executionContext);
+        RecordCursor slaveCursor = null;
+        try {
+            slaveCursor = slaveFactory.getCursor(executionContext);
+            if (masterCursor.hasNext()) {
+                cursor.of(masterCursor, slaveCursor, executionContext);
+                return cursor;
+            }
+            slaveCursor = Misc.free(slaveCursor);
+            Misc.free(masterCursor);
+        } catch (Throwable ex) {
+            Misc.free(masterCursor);
+            Misc.free(slaveCursor);
+            throw ex;
+        }
+
+        return EmptyTableRecordCursor.INSTANCE;
+    }
+
+    @Override
+    public boolean recordCursorSupportsRandomAccess() {
+        return false;
+    }
+
+    @Override
+    public boolean supportsUpdateRowId(CharSequence tableName) {
+        return masterFactory.supportsUpdateRowId(tableName);
+    }
+
+    @Override
+    protected void _close() {
+        ((JoinRecordMetadata) getMetadata()).close();
+        masterFactory.close();
+        slaveFactory.close();
+        filter.close();
+    }
+
+    private static class NestedLoopLeftRecordCursor extends AbstractJoinCursor {
+        private final Function filter;
+        private final OuterJoinRecord record;
+        private boolean hasMore;
+        private boolean isMatch;
+
+        public NestedLoopLeftRecordCursor(int columnSplit, Function filter, Record nullRecord) {
+            super(columnSplit);
+            this.record = new OuterJoinRecord(columnSplit, nullRecord);
+            this.filter = filter;
+            this.isMatch = false;
+            this.hasMore = true;
+        }
+
+        @Override
+        public Record getRecord() {
+            return record;
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (!hasMore) {
+                return false;
+            }
+
+            do {
+                while (slaveCursor.hasNext()) {
+                    if (filter.getBool(record)) {
+                        isMatch = true;
+                        return true;
+                    }
+                }
+
+                if (!isMatch) {
+                    isMatch = true;
+                    record.hasSlave(false);
+                    return true;
+                } else {
+                    isMatch = false;
+                    slaveCursor.toTop();
+                    record.hasSlave(true);
+                }
+            } while (masterCursor.hasNext());
+
+            hasMore = false;
+            return false;
+        }
+
+        @Override
+        public long size() {
+            return -1L;
+        }
+
+        @Override
+        public void toTop() {
+            masterCursor.toTop();
+            masterCursor.hasNext();
+            slaveCursor.toTop();
+            filter.toTop();
+            isMatch = false;
+            hasMore = true;
+            record.hasSlave(true);
+        }
```

Comment:
`fiter.toTop();`

Reply:
Do you mean toTop() method ? 
It's there already: ```java
            slaveCursor.toTop();
            filter.toTop();
            isMatch = false;
```

Reply:
yes, there is `filter.toTop()` already